crates/ra_assists/src/handlers/fill_match_arms.rs

```diff
@@ -104,6 +126,22 @@ fn resolve_enum_def(sema: &Semantics<RootDatabase>, expr: &ast::Expr) -> Option<
     })
 }
 
+fn resolve_tuple_of_enum_def(
```

Comment:
`resolve_enum_def` contains `autoderef` invocation.
This helps to detect more enum-related cases and might be a cool thing to support in `resolve_tuple_of_enum_def` too, now or later in some other PR.

Currently this code
```rust
let ref_opt = &Some(22);
match (ref_opt, ref_opt) {
}
```

will invoke the `panic!("handle the case of tuple containing non-enum"),` line and fail since we don't recognise the enum behind the reference.

This all kind of hints that we might want to approach the resolution somehow more unified, but that is maybe an another task outside of this PR's scope.

Reply:
Thanks for pointing this out. I've added the `autoderef` to `resolve_tuple_of_enum_def`, but I agree it does feel like there should be some way to reduce some of the duplication here.